File: C++/C++ 기초/기초문법/01_hello_world.cpp
/*
 * 01_hello_world.cpp
 * 
 * C++ 입출력과 네임스페이스 기초
 * - std::cout, std::cin을 이용한 입출력
 * - 네임스페이스(namespace) 개념과 사용법
 */

#include <iostream>  // 입출력 스트림 라이브러리
#include <string>     // 문자열 클래스

// 방법 1: 네임스페이스 전체 사용 (권장하지 않음)
// using namespace std;

// 방법 2: 특정 요소만 사용 (권장)
using std::cout;
using std::cin;
using std::endl;
using std::string;

int main() {
    // ========== 출력 (std::cout) ==========
    
    // 기본 출력
    cout << "Hello, World!" << endl;
    
    // 여러 값 출력
    int age = 25;
    string name = "홍길동";
    cout << "이름: " << name << ", 나이: " << age << endl;
    
    // 숫자 출력
    int num1 = 10, num2 = 20;
    cout << num1 << " + " << num2 << " = " << (num1 + num2) << endl;
    
    // ========== 입력 (std::cin) ==========
    
    int userAge;
    string userName;
    
    cout << "이름을 입력하세요: ";
    cin >> userName;  // 공백 전까지만 입력받음
    
    cout << "나이를 입력하세요: ";
    cin >> userAge;
    
    cout << "입력하신 정보: " << userName << ", " << userAge << "세" << endl;
    
    // ========== 공백 포함 문자열 입력 ==========
    
    // cin >> 는 공백을 만나면 입력이 끝남
    // 공백 포함 문자열을 입력받으려면 getline() 사용
    
    cin.ignore();  // 이전 입력의 개행 문자 제거
    
    string fullName;
    cout << "전체 이름을 입력하세요 (공백 포함): ";
    std::getline(cin, fullName);  // std:: 네임스페이스 명시
    cout << "전체 이름: " << fullName << endl;
    
    // ========== 네임스페이스 사용 방법 비교 ==========
    
    // 방법 1: 네임스페이스 전체 지정
    std::cout << "std:: 네임스페이스 직접 사용" << std::endl;
    
    // 방법 2: using 선언 사용 (위에서 선언함)
    cout << "using 선언으로 간단히 사용" << endl;
    
    // ========== 여러 데이터 타입 입출력 ==========
    
    int integer;
    float decimal;
    char character;
    bool boolean;
    
    cout << "\n정수 입력: ";
    cin >> integer;
    
    cout << "실수 입력: ";
    cin >> decimal;
    
    cout << "문자 입력: ";
    cin >> character;
    
    cout << "불리언 입력 (0 또는 1): ";
    cin >> boolean;
    
    cout << "\n입력된 값들:" << endl;
    cout << "정수: " << integer << endl;
    cout << "실수: " << decimal << endl;
    cout << "문자: " << character << endl;
    cout << "불리언: " << std::boolalpha << boolean << endl;  // true/false로 출력
    
    return 0;
}

/*
 * 네임스페이스 (Namespace)란?
 * 
 * 1. 목적: 이름 충돌 방지
 *    - 같은 이름의 함수나 변수를 다른 네임스페이스에서 사용 가능
 * 
 * 2. 사용 방법:
 *    - std::cout: 네임스페이스 직접 지정 (가장 명확)
 *    - using std::cout: 특정 요소만 가져오기 (권장)
 *    - using namespace std: 전체 네임스페이스 가져오기 (비권장)
 * 
 * 3. 왜 using namespace std를 피해야 하나?
 *    - 이름 충돌 가능성 증가
 *    - 코드의 가독성 저하 (어디서 온 것인지 불명확)
 *    - 대규모 프로젝트에서 문제 발생 가능
 * 
 * 4. 입출력 스트림:
 *    - cout: 표준 출력 (console output)
 *    - cin: 표준 입력 (console input)
 *    - endl: 줄바꿈 + 버퍼 플러시
 *    - << : 출력 연산자 (insertion operator)
 *    - >> : 입력 연산자 (extraction operator)
 */

File: Java/Java 기초/기초/01_hello_world.java
/*
 * 01_hello_world.java
 * 
 * Java 입출력과 클래스 기초
 * - System.out.println을 이용한 출력
 * - System.in, Scanner를 이용한 입력
 * - 클래스와 main 메서드 구조
 */

// 클래스 선언
public class HelloWorld {
    // main 메서드: 프로그램의 진입점
    public static void main(String[] args) {
        // ========== 출력 (System.out) ==========
        
        // println: 출력 후 줄바꿈
        System.out.println("Hello, World!");
        
        // print: 출력 후 줄바꿈 없음
        System.out.print("안녕하세요, ");
        System.out.print("Java입니다!");
        System.out.println(); // 줄바꿈만
        
        // printf: 포맷 지정 출력
        String name = "홍길동";
        int age = 25;
        System.out.printf("이름: %s, 나이: %d\n", name, age);
        
        // 여러 값 출력
        System.out.println("이름: " + name + ", 나이: " + age);
        
        // ========== 입력 (Scanner 사용) ==========
        
        // Scanner 클래스 import 필요 (파일 상단에 추가)
        // import java.util.Scanner;
        
        java.util.Scanner scanner = new java.util.Scanner(System.in);
        
        System.out.print("이름을 입력하세요: ");
        String userName = scanner.next(); // 공백 전까지만 입력받음
        
        System.out.print("나이를 입력하세요: ");
        int userAge = scanner.nextInt();
        
        System.out.println("입력하신 정보: " + userName + ", " + userAge + "세");
        
        // ========== 공백 포함 문자열 입력 ==========
        
        scanner.nextLine(); // 이전 입력의 개행 문자 제거
        
        System.out.print("전체 이름을 입력하세요 (공백 포함): ");
        String fullName = scanner.nextLine(); // 공백 포함 문자열 입력
        System.out.println("전체 이름: " + fullName);
        
        // ========== 여러 데이터 타입 입출력 ==========
        
        System.out.print("\n정수 입력: ");
        int integer = scanner.nextInt();
        
        System.out.print("실수 입력: ");
        double decimal = scanner.nextDouble();
        
        System.out.print("문자 입력: ");
        char character = scanner.next().charAt(0); // 첫 번째 문자만
        
        System.out.print("불리언 입력 (true 또는 false): ");
        boolean bool = scanner.nextBoolean();
        
        System.out.println("\n입력된 값들:");
        System.out.println("정수: " + integer);
        System.out.println("실수: " + decimal);
        System.out.println("문자: " + character);
        System.out.println("불리언: " + bool);
        
        scanner.close(); // Scanner 닫기
    }
}

/*
 * Java 프로그램 구조 핵심 정리
 * 
 * 1. 클래스 선언
 *    public class 클래스명 {
 *        // 클래스 내용
 *    }
 * 
 * 2. main 메서드
 *    - 프로그램의 진입점
 *    - 반드시 public static void main(String[] args) 형식
 *    - String[] args: 명령줄 인자
 * 
 * 3. 출력 방법
 *    - System.out.println(): 출력 후 줄바꿈
 *    - System.out.print(): 출력 후 줄바꿈 없음
 *    - System.out.printf(): 포맷 지정 출력 (C 스타일)
 * 
 * 4. 입력 방법
 *    - Scanner 클래스 사용
 *    - import java.util.Scanner; 필요
 *    - scanner.next(): 단어 입력 (공백 전까지)
 *    - scanner.nextLine(): 줄 전체 입력 (공백 포함)
 *    - scanner.nextInt(): 정수 입력
 *    - scanner.nextDouble(): 실수 입력
 *    - scanner.nextBoolean(): 불리언 입력
 * 
 * 5. 문자열 연결
 *    - + 연산자로 문자열 연결
 *    - "문자열" + 변수 + "문자열"
 * 
 * 6. 주석
 *    - // : 한 줄 주석
 *    - /* */ : 여러 줄 주석
 *    - /** */ : 문서 주석 (JavaDoc)
 */

